### PR TITLE
CoreML: Disable 1D ML Program matmul due to bug in coreml

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Disable using CoreML ML Program for a matmul where one of the inputs is 1D as the CoreML implementation appears to be broken. See https://github.com/apple/coremltools/issues/2263

Add some debugging notes. 

### Motivation and Context

Fix failing test on macos-14.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/builders/model_builder.cc [hunk 1]
```diff
@@ -906,6 +906,7 @@ Status ModelBuilder::SaveModel() {
 
 #if defined(COREML_ENABLE_MLPROGRAM)
   if (create_ml_program_) {
+    // we need to jump through some hurdles to get the model path the ML Program load wants.
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/model_builder.cc:909]: very minor nit: "jump through some hoops"

**Comment by edgchen1** [comment]
could you also try enabling the CoreML EP in the MacOS 14 CI build?

https://github.com/microsoft/onnxruntime/blob/8f738d8e9f03baa8958b22eb5a50d723f1ddbad4/.github/workflows/mac.yml#L50-L60

File: onnxruntime/core/providers/coreml/builders/model_builder.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/model_builder.cc:909]: you clearly haven't seen me hurdle... 🤣

**Review by snnn — DISMISSED** [review]

**Review by edgchen1 — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Pull request merged**